src/fileformat/pfsinrgbe.1
.TH "pfsinrgbe" 1
.SH NAME
pfsinrgbe \- Load images or frames in Radiance RGBE format
.SH SYNOPSIS
.B pfsinrgbe
[--linear] (<file> [--frames <range>] [--skip-missing])  [<file>...]

.SH DESCRIPTION
.I pfsinrgbe
command loads images in Radiance RGBE format and writes \fIpfs\fR
stream to the Standard Output. The \fIpfs\fR stream is usually piped
to another program for further processing. To detect the format
automatically based on the extension, use \fIpfsin\fR command.
.PP
Since RGBE is an HDR format, 'LUMINANCE' tag is set to 'RELATIVE'
(relative linear luminance values).
.PP
To read images from standard input use a single dash '-' instead of
filename. The images are read until EOF is reached.
.PP
Each file can contain a \%%d pattern, which is substituted with frame
numbers. The pattern has the same syntax as C
.I printf
command. For example, you can use \%%04d to make the frame number
four digit with proceedings zeros. You can select the frames using the
following options (the options must be always given after the file
name):
.TP
.B \--frames <range>
Range is given in mathlab / octave format:

.B "startframe:step:endframe"

Frame numbers start with
.B "startframe"
(default 0), are increased by
.B "step"
(default 1) and stop at
.B "endframe"
You can skip one of those values, for example
.I "1:100"
for frames 1,2,...,100 and
.I 0:2:
for frame 0,2,4,... up to the last file that exists.

.TP
.B \--skip-missing
Skip up to ten frames in a row if corresponding files are
missing. Otherwise the program stops reading sequence at the first
file that does not exists. This switch does not apply to the first
frame in a sequence. This switch can be useful if there is a rendered
animation where some of the frame has not been generated.

.TP
.B \--linear
Ignored for compatibility with \fIpfsinppm\fR.

.SH EXAMPLES
.TP
pfsinrgbe intro\%%d.hdr frame\%%d.hdr | ...

Concatenate two animations into one and pass it to pipe for further
processing.

.SH IMPORTANT NOTE

There is an incompatibility between viewers of RGBE format in terms of
absolute luminance values. The ximage viewer from Radiance package
expects radiance values in RGBE files and uses the WHITE_EFFICACY
constant equal to 179.0f to convert from radiance to luminance. This
in practice means that the values displayed with pfsview maybe higher
than expected. To compensate for this incompatibility we suggest to
use \fIpfsabsolute\fR command:

pfsinrgbe measured_luminance.hdr | pfsabsolute 1 179 | pfsview

This is not necessary when RGBE files are saved using
\fIpfsoutrgbe\fR, Radiance package, mkhdr or Photospheare. This is
also not relevant when only relative absolute values are considered.


.SH BUGS
Please report bugs and comments on implementation to 
the discussion group http://groups.google.com/group/pfstools
.SH "SEE ALSO"
.BR pfsin (1),
.BR pfsout (1)




ChangeLog
pfstools 1.8.5 <21.10.2011>
	* fixed: pfsintiff normalizes 16-bit files to 0-1 range to maintain compatibility with pfscalibration and the rest of pfstools
	* fixed: when reading TIFF files, pfsin falls back to pfsinimgmagick if pfsintiff not found
	* fixed: pfsout tries first pfsoutimgmagic when writing TIFF images

pfstools 1.8.4 <20.05.2011>
	* fixed: SegFault from pfsview when switching channels
	* fixed: quoted arguments in pfsin/pfsout/pfsv/pfsindcraw/pfsinmulti (fixes bug 3279342)

pfstools 1.8.3 <23.03.2011>
	* fixed: pfsdisplayfunction properly interpolates display LUT
	* fixed: bug #3080304 - configure uses non-portable test(1) syntax
	* fixed: bug #1766263 - pfsinhdrgen fails silently on pfsin failure. - partially
	* fixed: pfsview the entire code converted from qt3 to qt4 (fixes bug #3127946)
	* fixed: "Fit window to content" works again in pfsview
	* added: window icon in pfsview
	* added: improved image file saving in pfsview

pfstools 1.8.2 <18.06.2010>
	* fixed: compiler incompatibility issue in pfspanoramic.cpp
	* fixed: bug #2953028 "gcc 4.4.1 const cast error"
	* fixed: spaces were improperly escaped in pfsin / pfsout (thanks to Timo)
	* added: new improved hdrhtml template from hdrlabs.com (thanks to Christian)

pfstools 1.8.1 <01.06.2009>
	 * fixed: hdrhtml JavaScript code for XHTML pages
	 * fixed: hdrhtml '+' and '-' keys work now the same as in pfsview (were swapped)
	 * fixed: added missing includes
	 * fixed: matlab/pfs_shell.m removes matlab paths from LD_LIBRARY_PATH
	 * fixed: uninitialized color clipping mode in pfsview (tracker 2790026)
	 * fixed: pfsview, "preserve bri & hue" color clipping handles lower limit correcly now
	 * fixed: pfsinppm does not report EOF error when reading from stdin
	 * fixed: compilation on Mac OS X 10.5.7 with Fink
	 * added: make install creates application bundle for pfsview

pfstools 1.8 <23.02.2009>
	 * added: pfsouthdrhtml for generating web pages with an HDR viewer
	 * added: matlab functions: pfs_write_image, pfs_read_image
	 * added: matlab interface can now save multi-channel (>3) images
	 * fixed: cleaned up and improved matlab documentation, added Content.m
	 * fixed: problem with inheriting matlab's LD_LIBRARY_PATH when running pfs commands
	 * fixed: improved search for NETPBM header files in the configure script
	 * fixed: pfssize keeps aspect ratio when resizing many images with --min/max/x/y

pfstools 1.7.0 <22.10.2008>
	 * added: pfsingdal - reader of geospatial data formats using GDAL (Geospatial Data Abstraction Library) (thanks to Martin Lambers)
	 * added: pfsdisplayfunction command for conversion between display luminance and pixel values
	 * added: pfsinimgmagick and pfsinppm add a WHITE_Y tag to frames (needed by some tone mapping operators)
	 * pfsview: automatically switch to linear mapping and LDR range for display-referred images
	 * fixed: pfsview correctly shows pixel position x=0
	 * added: inverse gamma correction in pfsgamma
	 * added: pfsgamma sets propertly LUMINANCE tag and displays warnings on improper input images
	 * added: configure script displays information and commands that will not be compiled because of missing dependencies
	 * fixed: some warning messages due to stricter syntax in g++ 4.2.1
	 * bugfix: removed depreciated matlab command pfsread - caused compilation problems under Windows
	 * added: pfs_test_shell for quick testing for common matlab setup problems
	 * added: pfsview shows in the status bar current exposure for the dynamic range window (relative to Y=1)
	 * added: configure should automatically find the include dir for netpbm	
	 * fixed: handling of >8bit files is unified for all commands; new tag ('BITDEPTH') is used to store information about the bit-depth recission; pfsoutimagemagic and pfsoutppm support a new option --bit-depth
         * fixed: removed --linear option from pfsoutppm, pfsoutimagemagick, pfsouttiff and changed default behavior, which was too confusing and error-prone. Images will never be gamma-corrected (transformed to the sRGB) unless it is explicitly enforced with a new option '--srgb'.
	
pfstools 1.6.5 <06.05.2008>
	 * fixed: matlab interface cleanup: pfsfclose.cpp pfsfgets.cpp pfsfputs.cpp pfspopen2.cpp pfssend.cpp moved to another project as they do not belong to pfstools
	 * fixed: matlab Makefile automatically detects correct mex-file extension (thanks to Neil Alldrin)
	 * fixed: matlab - pfs_shell under unix adds bash as the default shell for executing pfs commands
	 * fixed: matlab interface operates now on single precission floats, thus making most operations faster (thanks to Neil Alldrin)
	 * added: matlab -> pfsview function shows now matrix names
	 * Debian patch ported: changes required to switch to octave3.0 (thanks to Thomas Weber)
	 * Debian patch ported: Fixed the usage of dcraw(1)'s -m command line option (thanks to Sebastian Harl)
	
pfstools 1.6.4 <01.01.2007>
	 * fixed: buffer overflow vulnerability in rgbeio.cpp (thanks to Stefan)
	 * fixed: compilation issues with gcc-4.3 - missing includes (thanks to Sebastian Harl for the Debian patch) 
	 
pfstools 1.6.3 <05.12.2007>
	 * fixed: pfsinopenexr can now read files that has data window < display window
	 * fixed: pfs library can handle channel names up to 32 characters (was 8)
	 * updated: pfs specification - max string lengths and format of custom channel names
	 * security fix: fscanf in rgbeio.cpp (thanks to Stefan and Ludwig)
	 * bugfix: pfsview - segfault when switching channels
	 * fixed: pfsview - color readout in the status line refreshed when new channel or frame loaded
	 * added: matlab/pfs_write_luminance.m
	 * bugfix: rgbeio - header read properly when the first byte is 0x20 (thanks to Axel)
	 * bugfix: rgbeio - fix misinterpreted rle-compresses lines (thanks to Axel)

pfstools 1.6.2 <04.07.2007>
         * matlab: pfsview can now display 2D cell arrays
	 * pfs library: quite serious bug in sRGB transforms fixed
	 * added: check for GLUT library (unix only)

src/fileformat/pfsinrgbe.cpp
/**
 * @brief Read files in Radiance's RGBE format
 * 
 * This file is a part of PFSTOOLS package.
 * ---------------------------------------------------------------------- 
 * Copyright (C) 2003,2004 Rafal Mantiuk and Grzegorz Krawczyk
 * 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * ---------------------------------------------------------------------- 
 * 
 * @author Grzegorz Krawczyk, <krawczyk@mpi-sb.mpg.de>
 * @author Rafal Mantiuk, <mantiuk@mpi-sb.mpg.de>
 *
 * $Id: pfsinrgbe.cpp,v 1.3 2008/01/01 13:01:21 rafm Exp $
 */

#include <config.h>

#include <cstdlib>

#include <iostream>

#include <getopt.h>
#include <pfs.h>

#include "rgbeio.h"

#define PROG_NAME "pfsinrgbe"

class QuietException 
{
};

void printHelp()
{
  std::cerr << PROG_NAME " [--verbose] [--help]" << std::endl
            << "See man page for more information." << std::endl;
}


void readFrames( int argc, char* argv[] )
{
  pfs::DOMIO pfsio;

  bool verbose = false;
  bool opt_linear=false;
  
  // Parse command line parameters
  static struct option cmdLineOptions[] = {
    { "help", no_argument, NULL, 'h' },
    { "verbose", no_argument, NULL, 'v' },
    { "linear", no_argument, NULL, 'l' },
    { NULL, 0, NULL, 0 }
  };
  static const char optstring[] = "lhv";
    
  pfs::FrameFileIterator it( argc, argv, "rb", NULL, stdin,
    optstring, cmdLineOptions );
    
  int optionIndex = 0;
  while( 1 ) {
    int c = getopt_long (argc, argv, optstring, cmdLineOptions, &optionIndex);
    if( c == -1 ) break;
    switch( c ) {
    case 'h':
      printHelp();
      throw QuietException();
    case 'v':
      verbose = true;
      break;
    case 'l':
      std::cerr << PROG_NAME << " warning: linearize option ignored for an HDR input!"
                << std::endl;
      break;
    case '?':
      throw QuietException();
    case ':':
      throw QuietException();
    }
  }

  while( true ) {
    pfs::FrameFile ff = it.getNextFrameFile();
    if( ff.fh == NULL ) break; // No more frames

    RGBEReader reader( ff.fh );

    VERBOSE_STR << "reading file (linear) '" << ff.fileName << "'" << std::endl;

    pfs::Frame *frame = pfsio.createFrame( reader.getWidth(),
					 reader.getHeight() );
    pfs::Channel *X, *Y, *Z;
    frame->createXYZChannels( X, Y, Z );

    //Store RGB data temporarily in XYZ channels
    reader.readImage( X, Y, Z );
    pfs::transformColorSpace( pfs::CS_RGB, X, Y, Z, pfs::CS_XYZ, X, Y, Z );

    frame->getTags()->setString("LUMINANCE", "RELATIVE");
    const char *fileNameTag = strcmp( "-", ff.fileName )==0 ? "stdin" : ff.fileName;
    frame->getTags()->setString( "FILE_NAME", fileNameTag );

    pfsio.writeFrame( frame, stdout );
    pfsio.freeFrame( frame );

    it.closeFrameFile( ff );
  }
}


int main( int argc, char* argv[] )
{
  try {
    readFrames( argc, argv );
  }
  catch( pfs::Exception ex ) {
    fprintf( stderr, PROG_NAME " error: %s\n", ex.getMessage() );
    return EXIT_FAILURE;
  }        
  catch( QuietException ) {
    return EXIT_FAILURE;
  }        
  return EXIT_SUCCESS;
}
